ci/build_matrix.py

```diff
@@ -0,0 +1,303 @@
+                "image_name": self.image_name,
+                "benchmark": self.benchmark,
+                "env_var": self.env_var,
+                "tag_prefix": f"{arch}-",
+                "arch": arch,
+                "changed": self.changed,
+            }
+
+
+class MatrixBuilder:
+    """
+    Builder for the GHA Jobs Matrix.
+
+    Parameters
+    ----------
+    archs : iterable of str
+        List of architectures to build against. Will create a matrix entry for each architecture for each
+        Dockerfile
+    bones : iterable of str
+        List of regex strings to match paths against to determine if the path is a snafu "bone".
+    upstream_branch : str
+        Upstream branch to compare changes to, in order to determine the value of "changed".
+    dockerfile_set : set of str
+        Set of dockerfiles within the snafu repository
+    changed_set : set of str
+        Set of changed files within the snafu repository
+    """
+
+    def __init__(
+        self,
+        archs: Iterable[str],
+        bones: Iterable[str],
+        upstream_branch: str,
+        dockerfile_set: Set[str],
+        changed_set: Set[str],
+    ):
+        """Contsruct the matrix builder."""
+
+        self.archs = archs
+        self.bones = bones
+        self.upstream_branch = upstream_branch
+        self.dockerfile_set = dockerfile_set
+        self.changed_set = changed_set
+        self.matrix: Dict[str, List[Dict[str, str]]] = dict()
+
+        self.reset()
+
+    def reset(self):
+        """Reset the matrix to empty starting point."""
+        self.matrix = {"include": []}
+
+    def add_entry(self, entry: MatrixEntry):
+        """Add the given MatrixEntry into the jobs matrix."""
+
+        for json_dict in entry.as_json():
+            self.matrix["include"].append(json_dict)
+
+    def bones_changed(self) -> bool:
+        """Return True if a bone has is found in the changed set."""
+
+        for bone in self.bones:
+            bone_regex = re.compile(bone)
+            for changed in self.changed_set:
+                if bone_regex.search(changed) is not None:
+                    return True
+        return False
+
+    def benchmark_changed(self, dockerfile: str) -> bool:
+        """Return True if the given dockerfile's benchmark has changed."""
+
+        dockerfile_dir = pathlib.Path(dockerfile).parent
+        for changed in self.changed_set:
+            try:
+                pathlib.Path(changed).relative_to(dockerfile_dir)
+            except ValueError:
+                pass
+            else:
+                return True
+        return False
+
+    def build(self):
+        """Build the GHA jobs matrix."""
+
+        bones_changed = self.bones_changed()
+        for dockerfile in self.dockerfile_set:
+            changed = bones_changed or self.benchmark_changed(dockerfile)
+            entry = MatrixEntry.new(dockerfile=dockerfile, archs=self.archs, changed=changed,)
+            self.add_entry(entry)
```

Comment:
Rather than adding an entry for changed/not-changed benchmarks (AKA dockerfiles), why don't we generate a list with only changed benchmarks?, that way we could skip the parameter changed and generate a more user friendly output :).

```python
            if bones_changed or self.benchmark_changed(dockerfile):
                entry = MatrixEntry.new(dockerfile=dockerfile, archs=self.archs)
                self.add_entry(entry)
```

Reply:
Implemented! I added a new option to the CLI `--changed-only` which will only output changed Dockerfiles. I want to keep in the ability to output Dockerfiles that haven't changed as well, in order to support a manual re-build of all images for testing/debug purposes.